opds_import.py

```diff
@@ -663,6 +662,90 @@ def summary_to_linkdata(detail):
         return kwargs_meta
 
     @classmethod
+    def coveragefailures_from_messages(cls, data_source, parser, feed_tag):
+        """Extract CoverageFailure objects from a parsed OPDS document. This
+        allows us to determine the fate of books which could not
+        become <entry> tags.
+        """
+        path = '/atom:feed/simplified:message'
+        for message_tag in parser._xpath(feed_tag, path):
+            failure = cls.coveragefailure_from_message(
+                data_source, parser, message_tag
+            )
+            if failure:
+                yield failure
+
+    @classmethod
+    def coveragefailure_from_message(cls, data_source, parser, message_tag):
+        """Turn a <simplified:message> tag into a CoverageFailure."""
+       
+        # First thing to do is determine which Identifier we're
+        # talking about. If we can't do that, we can't create a
+        # CoverageFailure object.
+        identifier_tag = parser._xpath1(message_tag, 'atom:id')
+        if identifier_tag is None:
+            urn = None
+        else:
+            urn = identifier_tag.text
+
+        # What will the CoverageFailure say is the problem?
+        description_tag = parser._xpath1(message_tag, 'schema:description')
+        if description_tag is None:
+            description = ''
+        else:
+            description = description_tag.text
+        
+        # What status code is associated with the error?
+        status_code_tag = parser._xpath1(message_tag, 'simplified:status_code')
+        if status_code_tag is None:
+            status_code = None
+        else:
+            status_code = status_code_tag.text
+
+        return cls.coveragefailure_from_message_data(
+            data_source, urn, status_code, description
+        )
+
+    @classmethod
+    def coveragefailure_from_message_data(
```

Comment:
This is the method that gets heavily tested; the methods above are only tested via higher-level methods that call them.